src/storage/txn/latch.rs

```diff
@@ -0,0 +1,137 @@
+// Copyright 2016 PingCAP, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+
+use std::collections::{VecDeque, BTreeSet};
+use std::hash::{Hash, SipHasher, Hasher};
+
+struct Latch {
+    // store waiting commands
+    pub waiting: VecDeque<u64>,
+
+    // use to check existed
+    pub set: BTreeSet<u64>,
+}
+
+impl Latch {
+    pub fn new() -> Latch {
+        Latch {
+            waiting: VecDeque::new(),
+            set: BTreeSet::new(),
+        }
+    }
+}
+
+pub struct Latches {
+    slots: Vec<Latch>,
+    size: usize,
+}
+
+impl Latches {
+    pub fn new(size: usize) -> Latches {
+        Latches {
+            slots: (0..size).map(|_|  Latch::new()).collect(),
+            size: size,
+        }
+    }
+
+    pub fn calc_slots<H>(&self, keys: &[H]) -> Vec<usize>
+        where H: Hash {
+        // prevent from deadlock, so we sort and deduplicate the index
+        let mut slots: Vec<usize> = keys.iter().map(|x| self.calc_slot(x)).collect();
+        slots.sort();
+        slots.dedup();
+        slots
+    }
+
+    pub fn acquire(&mut self, slots: &[usize], who: u64) -> usize {
+        let mut acquired_count: usize = 0;
+        for i in slots {
+            let latch = &mut self.slots.get_mut(*i).unwrap();
+
+            let mut front: Option<u64> = None;
+            if let Some(cid) = latch.waiting.front() {
+                front = Some(*cid);
+            }
+
+            match front {
+                Some(cid) => {
+                    if cid == who {
+                        acquired_count += 1;
+                    } else {
+                        if !latch.set.contains(&who) {
```

Comment:
you can use insert directly here and check the return to see whether who existed in set before.